File: backend/src/main/java/dev/coldhands/pair/stairs/backend/infrastructure/web/controller/GenericErrorHandler.java
package dev.coldhands.pair.stairs.backend.infrastructure.web.controller;

import dev.coldhands.pair.stairs.backend.infrastructure.web.dto.ErrorDto;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;

@ControllerAdvice
public class GenericErrorHandler {

    @ExceptionHandler(MethodArgumentNotValidException.class)
    public ResponseEntity<ErrorDto> handleError(MethodArgumentNotValidException e) {
        final var errorCode = e.getBindingResult().getAllErrors().getFirst().getDefaultMessage();
        return ResponseEntity.badRequest().body(new ErrorDto(errorCode));
    }
}

File: backend/src/main/kotlin/dev/coldhands/pair/stairs/backend/infrastructure/web/controller/GenericErrorHandler.kt
package dev.coldhands.pair.stairs.backend.infrastructure.web.controller

import dev.coldhands.pair.stairs.backend.infrastructure.web.dto.LegacyErrorDto
import org.springframework.http.ResponseEntity
import org.springframework.http.ResponseEntity.badRequest
import org.springframework.http.converter.HttpMessageNotReadableException
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ExceptionHandler

@ControllerAdvice
class GenericErrorHandler {

    @ExceptionHandler(MethodArgumentNotValidException::class)
    fun handleError(e: MethodArgumentNotValidException): ResponseEntity<LegacyErrorDto> {
        val errorCode: String = e.bindingResult.allErrors.firstOrNull()?.defaultMessage
            ?: "UNKNOWN_VALIDATION_ERROR"
        return badRequest().body(LegacyErrorDto(errorCode))
    }

    @ExceptionHandler(HttpMessageNotReadableException::class)
    fun handleError(): ResponseEntity<LegacyErrorDto> {
        return badRequest().body(LegacyErrorDto("INVALID_REQUEST_BODY"))
    }
}